File: tourism/src/main/java/com/tourism/Booking.java
package com.tourism;

import com.google.inject.Inject;

/**
 * Клас для управління бронюванням (композиція з BookingItem та агрегація з BookingService)
 */
public class Booking {
    private BookingItem item; // композиція
    private PaymentService payment;
    private BookingService bookingService;

    /**
     * Конструктор з впровадженням залежності від сервісу збереження даних
     *
     * @param bookingService сервіс для збереження бронювання
     */
    @Inject
    public Booking(BookingService bookingService) {
        this.bookingService = bookingService;
    }

    /**
     * Створює позицію бронювання та сервіс оплати
     *
     * @param roomType тип номера
     * @param price ціна
     */
    public void initialize(String roomType, double price) {
        this.item = new BookingItem(roomType, price);
        this.payment = new PaymentService();
    }

    /**
     * Підтверджує бронювання та зберігає його в базу даних
     */
    public void confirm() {
        System.out.println("Підтвердження бронювання...");
        payment.processPayment(item.getPrice());
        // Збереження позиції бронювання в базу даних
        bookingService.saveBookingItem(this.item);
    }

    /**
     * Виводить інформацію про бронювання
     */
    public void showBooking() {
        System.out.println("Бронювання: " + item.getRoomType() + " | " + item.getPrice() + "$");
    }

    public BookingItem getItem() {
        return item;
    }
}

File: tourism-python/src/tourism/booking.py
from tourism.booking_item import BookingItem
from tourism.payment_service import PaymentService

class Booking:
    """
    Клас для управління бронюванням
    композиція: BookingItem
    агрегація: BookingService
    """

    def __init__(self):
        self.item = None
        self.payment = None
        self.booking_service = None  # буде injected через setter

    # === Setter-based DI (аналог @Inject setter) ===
    def set_booking_service(self, booking_service):
        self.booking_service = booking_service

    def initialize(self, room_type: str, price: float):
        self.item = BookingItem(room_type, price)
        self.payment = PaymentService()

    def confirm(self):
        print("Підтвердження бронювання...")
        self.payment.process_payment(self.item.price)
        self.booking_service.save_booking_item(self.item)

    def show_booking(self):
        print(f"Бронювання: {self.item.room_type} | {self.item.price}$")

    def get_item(self):
        return self.item
